### PR TITLE
GH-38432: [C++][Parquet] Trying to use original logic of BinaryBuilder

### LINKED ISSUE #38432: [C++] Parquet reading performance regressions
Looking at some of the Parquet read benchmarks we have (https://conbench.ursa.dev/c-benchmarks/file-read), there are two small regressions to be observed the last months.

For example, zooming in on that time for nyctaxi_2010 with snappy compression (https://conbench.ursa.dev/benchmark-results/0653775157e07b6980005bce6e59a2ad/):

image

And for fanniemae (https://conbench.ursa.dev/benchmark-results/0653774cc429733d8000b2db91fa63fa/); here only the first bump is noticeable:

image

There are two small bumps, which I think are caused by the following two PRs, in order of the bumps:

- https://github.com/apache/arrow/pull/14341 
  This PR added a new encoding option, but so my expectation is that it shouldn't have to slow down the default (the new encoding isn't used)? (cc @rok)
- https://github.com/apache/arrow/pull/37854
  While this change is mostly useful for cloud filesystems, and could be detrimental for fast local disk, the general assumption was that this effect for local disk would not be significant, so we could simply enable it for all filesystems. But based on this benchmark it seems the impact might actually not be fully insignificant? Should we consider making the default for `pre_buffer` depend on the type of filesystem?

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

Do some changes mentioned in https://github.com/apache/arrow/issues/38432

### What changes are included in this PR?

1. for `std::optional` argument, it might be used in loop, so separate it
2. Remove some `Prepare`.

### Are these changes tested?

Already has tests

### Are there any user-facing changes?

no

* Closes: #38432

### CODE REVIEW

**Comment by raulcd** [comment]
@ursabot please benchmark

**Comment by mapleFU** [comment]
@ursabot please benchmark

**Comment by jorisvandenbossche** [comment]
Looking at the benchmark run, there is a clear improvement in the affected `file-read` benchmarks. Asking for the comparison of the benchmarked commit with the PR base: https://conbench.ursa.dev/compare/runs/b9dce8c2b76f4c278256da7797cb932d...a810bbfce49c4c238686b2097866fc39/, and sorting by positive z score:

image

So this seems to solve the issue that is discussed in https://github.com/apache/arrow/issues/38432

**Comment by mapleFU** [comment]
I didn't find some stable improvement from local benchmark, profile and others

But maybe we can first move forward.

**Comment by mapleFU** [comment]
I didn't find some stable improvement from local benchmark, profile and others

But maybe we can first move forward.

**Comment by mapleFU** [comment]
@jorisvandenbossche @pitrou 

I've revert some BinaryHelper change. I don't know the reason because I don't have the regression workload. Maybe you can take a look

File: cpp/src/parquet/encoding.cc [hunk 1]
```diff
@@ -1205,16 +1205,22 @@ struct ArrowBinaryHelper<ByteArrayType> {
...
-                          std::optional<int64_t> estimated_remaining_data_length = {}) {
+  Status PrepareNextInput(int64_t next_value_length) {
+    if (ARROW_PREDICT_FALSE(!CanFit(next_value_length))) {
+      // This element would exceed the capacity of a chunk
+      return PushChunk();
+    }
+    return Status::OK();
+  }
+
+  Status PrepareNextInputWithEstimatedLength(int64_t next_value_length,
```
Comment (mapleFU) [inline cpp/src/parquet/encoding.cc:1216]: Split into two functions to make it fast and clear

File: cpp/src/parquet/encoding.cc [hunk 2]
```diff
@@ -1983,7 +1990,6 @@ class DictByteArrayDecoderImpl : public DictDecoderImpl<ByteArrayType>,
     int values_decoded = 0;
 
     ArrowBinaryHelper<ByteArrayType> helper(out, num_values);
-    RETURN_NOT_OK(helper.Prepare(len_));
```
Comment (mapleFU) [inline cpp/src/parquet/encoding.cc:1986]: Ditto

**Comment by pitrou** [comment]
Can you please explain clearly why you're doing these changes? It's not obvious at all that they are correct.

**Comment by mapleFU** [comment]
That's because:
1. https://github.com/apache/arrow/issues/38432 This issue mentioned a performance regression when using 14.0 and Find that DELTA_BYTE_ARRAY causing build arrow slower. After my change the regression disappear but I don't know why =_=
2. I think https://github.com/apache/arrow/issues/38577 also raise from same reason, but I don't know if is ok to fix it.

**Comment by mapleFU** [comment]
I've created a simpler patch for this.

**Pull request closed without merging**